tools/link.py

```diff
@@ -620,6 +622,12 @@ def phase_linker_setup(options, state, newargs):
   if settings.OPT_LEVEL >= 1:
     default_setting('ASSERTIONS', 0)
 
+  if settings.GZIP_EMBEDDINGS and not settings.SINGLE_FILE:
+    exit_with_error('GZIP_EMBEDDINGS is not currently meaningfull except in SINGLE_FILE mode')
+
+  if settings.SINGLE_FILE and settings.SHRINK_LEVEL >= 2:
+    settings.GZIP_EMBEDDINGS = 1
```

Comment:
This part seems a little unnecessary to me, and adds complexity. I agree it's a reasonable default but it seems better to have users set it manually.

Reply:
Why not do it by default in cases where we know for sure it will make the overall codesize smaller?

Reply:
I worry the user might notice the difference if they inspect the contents, which might be slightly confusing when debugging and changing optimization flags. But I guess it isn't a normally observable difference - no extra files or such - so I don't feel strongly.

Reply:
I think this is all an implementation detail. I can't imagine folks noticing if we change it.. other than the payload size changing.  Can you think of some way that somebody might depend on the absolute base64 encoded content in these interal blob urls?

Reply:
When debugging you might notice this, either by reading the code or by inspecting it as it executes. But I agree it's minor.

Reply:
But by that logic one could object to almost any internal change to how emscripten does just about anything, no?

Reply:
Sorry, I think I didn't explain myself well. There is a fairly big "discontinuity" that happens between `-Os` and `-Oz` with this change, and we have few such things atm. Any such discontinuity has a risk of being annoying during debugging - I've lost many hours to such things in the past.

By "big discontinuity" I mean that going from `-Oz` to `-Os` changes a big chunk of code (bundled fflate) and the entire algorithm used to encode. Imagine you are debugging a miscompile in `-Oz` but not in `-Os`, then this would be another difference you would need to consider; in particular debug loggings in places regarded to loading the wasm would need to be different. That isn't theoretical: debugging wasm2js builds has been an annoyance in the past because how the wasm is loaded is different vs a normal build (leading to `MAYBE_WASM2JS` mode which helped my many times).

Of course we have such big discontinuities elsewhere, e.g. `-O1` does not inline and `-O2` does, or a totally different register allocator might be used in `-O3`. And each of those differences can have a downside if the user expects a small and continuous difference between them. So it's a matter of degree and of cost/benefit.

But, again, I accept this is a minor matter and I don't strongly object to it, as I've said. The cost/benefit may be worth it here! :smile:

Reply:
Whether or not to use fflate probably shouldn't be decided based on `-Oz` vs `-Os` - if the goal is smaller code then for WASM file bigger than about 12kB it would be smaller.

The time it takes to decompress could be noticeable for very large files. I just put my biggest WASM file (3MB) through the fflate demo, and it compressed down to 934kB (31%) in 260ms. I haven't measured the actual decompression time, but I'd expect it to be the same order of magnitude.

Does anyone use SINGLE_FILE with 10MB+ WASM files? I'd suggest that GZIP_EMBEDDINGS be enabled by default with SINGLE_FILE, but be manually disableable, if anyone is concerned about that decompression time. WASM files bigger than 12-15kB should be compressed, but I don't know if the setting needs to actually be changed based on the file size, or if it can just have no effect if the file is smaller.

Reply:
I think we should opt into this for `-Os` and `-Oz` builds if (and only of) the resulting file (i.e. the compressed string + the decompression code) is smaller than the non-compressed version. 

For folks using `-Os` and `-Oz` we already know they prefer size over performacnce so that small performce hit due to decompression is expected.

In terms of the `"big discontinuity" I mean that going from -Oz to -Os`.. does your concert go away once I clarity that I indented this for both `-Oz` and `-Os`?

I think doing from `-O3` of `-Oz` (or `-Os`) we already have plenty of big differences so I don't see that as a problem.